## Alignment to reference genomes

From our previous work we have already generated BAM files for each of these samples ([Kleinmann-Ruiz et al. 2022](https://www.pnas.org/doi/abs/10.1073/pnas.2110614119) and [Bazzicalupo et al. 2023](https://onlinelibrary.wiley.com/doi/full/10.1111/eva.13570)), generated using the Canada lynx assembly as the reference genome ([mLynCan4_v1.p - GCA_007474595.1](https://www.ensembl.org/Lynx_canadensis/Info/Index); [Rhie et al., 2021](http://www.nature.com/articles/s41586-021-03451-0)).

Additionally I used sequences generated for the new Plan Nacional project. Libraries were generated by Luc√≠a and Laura and sequenced by novogene *check with them for additional details*

### Data Quality Control and Read Processing

I run [fastqc](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) to check the quality of rawreads from sequencing.

Seeing the warnings I receive from fastqc reports I need to process the rawreads to try to remove them.

I use [fastp](https://github.com/OpenGene/fastp) from [Chen et al. (2018)](https://academic.oup.com/bioinformatics/article/34/17/i884/5093234?login=true) with default settings plus the following flags:
```
--dont_overwrite (protect the existing files not to be overwritten by fastp)
--trim_poly_g (detect the polyG in read tails and trim them)
--length_required 30 (reads shorter than length_required will be discarded)
--correction (enable base correction in overlapped regions)
--detect_adapter_for_pe (enable adapter sequence auto-detection)
--thread 6 (worker thread number)
```

See [Lucia's github](https://github.com/luciamayorf/Data_preprocessing_alignment?tab=readme-ov-file#2-reads-trimming-and-quality-control) for how these scripts were run.

### Run alignments

I align the my samples to the [Lynx canadensis reference genome](https://www.ensembl.org/Lynx_canadensis/Info/Index).

A script will be run for each sample that will use [bwa v0.7.17](https://bio-bwa.sourceforge.net/bwa.shtml), [samtools v1.9](https://www.htslib.org/doc/samtools.html), [picard v2.25.5](https://broadinstitute.github.io/picard/) and [gatk v3.7-0](https://gatk.broadinstitute.org/hc/en-us) that will do the following:
- Align each of the sample's R1-R2 fastq pairs to the reference genome using BWA-MEM and Samtools view
- samtools sort to sort the reads in the bam file
- picardtools AddOrReplaceReadGroups to add read groups to the reads in the bam file
- samtools merge to merge the bam files from the multiple R1-R2 pairs of the sample if necessary
- picardtools MarkDuplicates to mark duplicate reads in the bam
- realign indels with GATK which comprises:
    - gatk RealignerTargetCreator to identify realign targets
    - gatk IndelRealigner to realign the indels
- samtools index to index the indel realigned bam for downstream analyses

Scripts that will run the alignment pipeline [for each sample](src/print_samples_in_config.py) in the [alignment configuration file](config/alignments/all_fastp_alignment.yml) are generated using the [run_alignment](src/congenomics_fastq_align-main/run_alignment.py) python script with the flag `--test`:
```
cd scripts/alignments

for sample in $(python ../../src/print_samples_in_config.py ../../config/alignments/all_fastp_alignment.yml)
 do
  echo "generating alignment script of $sample"
  python ../../src/congenomics_fastq_align-main/run_alignment.py --sample $sample --config ../../config/alignments/all_fastp_alignment.yml --test
done
```
The generated scripts were then submitted to the job queue on cesga ft3:
```
# sbatch all except lr
for sh in $(ls scripts/alignments/*.sh)
 do
  echo "sbatching $sh"
  sbatch $sh
done
```

### Alignment quality control

To check on the quality of alignments I run [QualiMap v2.2.1](http://qualimap.conesalab.org/) ([Okonechnikov et al. 2016](https://doi.org/10.1093/bioinformatics/btv566)) using the [run_qualimap_bam_out](src/alignments/run_qualimap_bam_out.sh) script.

```
# Sbatch QualiMap of all samples I want to check out
bams=$(ls /mnt/lustre/hsm/nlsas/notape/home/csic/ebd/jgl/lynx_genome/lynx_data/LyCaRef_bams/*er.bam | grep -E "ll|lp" | grep -E "ll_ca|lp_sm|ll_ya|ll_vl|ll_ki|ll_ur" | grep -vE "ca_0249|ca_0253")

for bam in ${bams[@]}
 do
  echo "sbatch qualimap of $bam"
  sbatch src/alignments/run_qualimap_bam_out.sh \
  ${bam} data/qualimap
done
```

Then I can summarize the results using [multiqc](https://multiqc.info/) ([Ewels et al. (2016)](https://academic.oup.com/bioinformatics/article/32/19/3047/2196507)).

```
cd data/qualimap
multiqc .
```
